lib/logstash/outputs/s3.rb

```diff
@@ -272,6 +272,7 @@ def close
     # This will block the shutdown until all upload are done or the use force quit.
     @file_repository.each_files do |file|
       upload_file(file)
+      clean_temporary_file(file) if file.size == 0
```

Comment:
> I once more thought and moved the removal logic in the plugin#close. In clean_temporary_file there are debugging log lines and LS core has also debugging log before closing the plugin as I remember.

Agreed, I think it makes sense to move this logic here, as it seems the `close` method is the only place that might left empty files undeleted.

I've a concern about the latest change. If I'm not mistaken, not all `@file_repository.each_files` will be empty files, meaning that they might be deleted by the `upload_file` method before reaching this size check (`file.size == 0`), raising an error.

I haven't tested it, but I think an option would be using the `File.zero?`, or moving this logic after the uploader was stopped.

```suggestion
      clean_temporary_file(file) if File.zero?(file.key)
```

I guess It wasn't caught by the added tests because the files are not being synchronously deleted and/or timing of executing the `:on_complete` callback.

Reply:
Unfortunately, we cannot use `File.zero` here. `file` is a `TemporaryFile` instance which carries internally file descriptor (FD). When calculating the size, it has an internal calculation strategy that aligns on FD or calls directly `File#size` if FD is closed.

> they might be deleted by the upload_file method before reaching this size

Your concern makes sense for the worst case scenario. Upload is an async operation here and there is no guarantee when it finishes before size check or later. So far, I have been testing multiple times didn't face but for the safety adding `Dir.exist?(file.temp_path)` would be better.

Reply:
Just to clarify, as I don't know this plugin well, Isn't the `temp_path` the same for all temporary files? If it's, I'd let it as it was, and just `rescue` failures, so other files on the loop would have the chance to get uploaded if one fail. for example:

``` ruby
      ...
      begin
        clean_temporary_file(file) if file.size == 0
      rescue => e
        @logger.debug("Failed removing temporary file", :path => file.path, :exception => e.class, :message => e.message)
      end
      ...
```

Otherwise, LGTM! :)

Reply:
> Just to clarify, as I don't know this plugin well, Isn't the temp_path the same for all temporary files?

This plugin has alot magics. No, every temp path is unique.